File: Neetcode/1DDP/houserobber2.py
# https://neetcode.io/problems/house-robber-ii
# https://leetcode.com/problems/house-robber-ii/

from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        def helper(arr):
            rob1 = rob2 = 0
            for n in arr:
                rob1, rob2 = rob2, max(rob1 + n, rob2)
            return rob2

        return max(nums[0], helper(nums[1:]), helper(nums[:-1]))

File: Neetcode/1DDP/houserobber2.cpp
#include <vector>
using namespace std;

class Solution {
 public:
  int helper(vector<int>& nums, int s, int e) {
    int temp, r1 = 0, r2 = 0;
    for (int i = s; i < e; ++i) {
      temp = max(r1 + nums[i], r2);
      r1 = r2;
      r2 = temp;
    }
    return r2;
  }

  int rob(vector<int>& nums) {
    int n = nums.size();
    if (n == 1) {
      return nums[0];
    }
    return max(helper(nums, 0, n - 1), helper(nums, 1, n));
  }
};